code/controllers/subsystem/mapping.dm

```diff
@@ -0,0 +1,114 @@
+
+
+/datum/subsystem/mapping
+	name       = "Map"
+	init_order = SS_INIT_MAP
+	flags      = SS_NO_FIRE
+
+	///All possible biomes in assoc list as type || instance
+	var/list/biomes = list()
+
+/datum/subsystem/mapping/New()
+	NEW_SS_GLOBAL(SSmapping)
+
+
+/datum/subsystem/mapping/Initialize(timeofday)
+	if (config.enable_roundstart_away_missions)
+		log_startup_progress("Attempting to generate an away mission...")
+		createRandomZlevel()
+
+	var/watch
+	if (!config.skip_fixedvault_generation)
+		watch = start_watch()
+		log_startup_progress("Placing fixed space structures...")
+		generate_fixedvaults()
+		log_startup_progress("Finished placing fixed structures in [stop_watch(watch)]s.")
+	else
+		log_startup_progress("Not generating fixed vaults - SKIP_VAULT_GENERATION found in config/config.txt")
+
+	if (!config.skip_vault_generation)
+		watch = start_watch()
+		log_startup_progress("Placing random space structures...")
+		generate_vaults()
+		generate_asteroid_secrets()
+		make_mining_asteroid_secrets() // loops 3 times
+		log_startup_progress("Finished placing structures in [stop_watch(watch)]s.")
+	else
+		log_startup_progress("Not generating vaults - SKIP_VAULT_GENERATION found in config/config.txt")
+
+	//hobo shack generation, one shack will spawn, 1/3 chance of two shacks
+	generate_hoboshack()
+	if (rand(1,3) == 3)
+		generate_hoboshack()
+
+	watch = start_watch()
+	for(var/datum/zLevel/z in map.zLevels)
+		var/watch_prim = start_watch()
+		z.post_mapload()
+		log_debug("Finished with zLevel [z.z] in [stop_watch(watch_prim)]s.", FALSE)
+	log_debug("Finished calling post on zLevels in [stop_watch(watch)]s.", FALSE)
+
+	watch = start_watch()
+	map.map_specific_init()
+	log_debug("Finished map-specific inits in [stop_watch(watch)]s.", FALSE)
+
+	spawn_map_pickspawners() //this is down here so that it calls after allll the vaults etc are done spawning - if in the future some pickspawners don't fire, it's because this needs moving
+
+	watch = start_watch()
+	initialize_biomes()
+	log_startup_progress("Finished initializing biomes in [stop_watch(watch)]s.")
+
+	..()
+
+/proc/generate_planet()//debug
+	return SSmapping.spawn_planetoid(/datum/planet_type/snow, /datum/map_element/mining_surprise/crashed_tradeship)
+
+///Initialize all biomes, assoc as type || instance
+/datum/subsystem/mapping/proc/initialize_biomes()
+	for(var/biome_path in subtypesof(/datum/biome))
+		var/datum/biome/biome_instance = new biome_path()
+		biomes[biome_path] += biome_instance
+
+/datum/subsystem/mapping/proc/spawn_planetoid(datum/planet_type/planet_datum, ruin_type)
+	var/datum/planetGenerator/mapgen = new planet_datum.mapgen
+	var/datum/map_element/mining_surprise/used_ruin = ispath(ruin_type) ? (new ruin_type) : ruin_type
+	message_admins("Generating turfs")
+	world.maxz += 1
+	map.addZLevel(new /datum/zLevel/away, world.maxz, TRUE, TRUE)
+	mapgen.generate_turfs(map.zLevels.len)
+
+	//DEBUG needs less hardcoding
+	for(var/x = 1,  x < 102, x++)
```

Comment:
You know this already but 102 need to be changed to something more modular.

Reply:
Now fixed